static/js/filter.js

```diff
@@ -82,7 +82,12 @@ function message_matches_search_term(message, operator, operand) {
         if (page_params.realm_is_zephyr_mirror_realm) {
             return zephyr_stream_name_match(message, operand);
         }
-        return (message.stream.toLowerCase() === operand);
+        var stream_id = stream_data.get_stream_id(operand);
+        if (!stream_id) {
+            return false;
+        }
+
+        return (message.stream_id === stream_id);
```

Comment:
I verified that the delete scenario you were worried about does indeed happen with this fix.  There are two ways to solve this:

* Make the stream-match check slightly more complicated by trying first to get a naive name==name match, and then if that fails, try to match at the id level using the code I just submitted.
* Have message fetching code tell stream_data about id/name mappings for deleted messages.

The first option is maybe a bit hacky, but it probably doesn't have any serious pitfalls.  It can have edge-case-y behavior for when the old message had a stream name that was later re-used by a new stream, I suppose.

The second option seems unappealing to me, as these skeleton subs can be kind of brittle.

We should probably discuss this on chat.